### DISCUSSION TITLE
Present MenuBarExtra window on app launch?

### DISCUSSION THREAD

**Discussion by andrewtheis** [orchetect/MenuBarExtraAccess#5] [Q&A] [answered]
I'm trying to automatically present the MenuBarExtra window on first app launch, so far this approach doesn't appear to work. Am I missing something obvious?

```swift
@main
struct MyApp: App {
    @NSApplicationDelegateAdaptor(AppDelegate.self) var appDelegate
    
    @ObservedObject var appState = AppState.shared
    
    var body: some Scene {
        MenuBarExtra("MyApp", systemImage: "music.note.tv") {
            MenuBarExtraView()
                .frame(minWidth: 800)
        }
        .menuBarExtraStyle(.window)
        .menuBarExtraAccess(isPresented: $appState.menuBarViewIsPresented)
    }
}

class AppState: ObservableObject {
    static let shared = AppState()
    @Published var menuBarViewIsPresented: Bool = false
}


class AppDelegate: NSObject, NSApplicationDelegate {
    func applicationDidFinishLaunching(_ notification: Notification) {
        DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
            AppState.shared.menuBarViewIsPresented = true
        }
    }
}
```

**Comment by orchetect** [comment, marked as answer]
Doing things on app launch with SwiftUI is tricky.

One issue may be that AppState is a global singleton when it should probably only be instanced in MyApp and passed by environment to child views.

It’s better if you are able to mutate SwiftUI state from within a SwiftUI context.

There is a slightly clunky way but it works:

```swift
@main struct MyApp: App {
    @Environment(\.scenePhase) private var scenePhase
    @State private var isLaunched: Bool = false
    @ObservedObject var appState = AppState()
   
    var body: some Scene {
        MenuBarExtra("MyApp", systemImage: "music.note.tv") {
            MenuBarExtraView()
                .frame(minWidth: 800)
        }
        .menuBarExtraStyle(.window)
        .menuBarExtraAccess(isPresented: $appState.menuBarViewIsPresented)
        .onChange(of: scenePhase) { newValue in
            switch newValue {
            case .active:
                if !isLaunched {
                    isLaunched = true
                    DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
                        appState.menuBarViewIsPresented = true
                    }
                }
            default: break
            }
        }
    }
}
```

**Comment by fffx** [comment]
`deadline: .now() + 0.1` doesn't work for me ,  I have to increase it to at least `+ 0.3` on my m1 MacBook Pro with Sonoma 14.6.1.

BTW,  we can also use`.commands` modifier, found this  from   https://stackoverflow.com/questions/76607396/show-a-window-on-startup-in-document-based-macos-app

**Reply by orchetect** [reply]
Anything involving an arbitrary delay of time is brittle. It's probably better to set it higher - like 1 or 2 seconds. Short enough that it's not really perceptible in the context of an app launching, but may help account for slower machines.

Eventually I would like to get to the bottom of a more reliable way to run code on app launch, and figure out if there is a way to get MenuBarExtraAccess to internally handle all of this if the presentation binding is defaulted to `true` on app launch. There may be a way to have it wait for the appropriate conditions post-launch to always guarantee the menu will be shown.
